File: extensions/standard-processors/processors/GenerateFlowFile.cpp
/**
 * @file GenerateFlowFile.cpp
 * GenerateFlowFile class implementation
 *
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "GenerateFlowFile.h"

#include <limits>
#include <memory>
#include <random>
#include <string>
#include <vector>

#include "minifi-cpp/core/ProcessContext.h"
#include "core/ProcessSession.h"
#include "core/Resource.h"
#include "utils/OptionalUtils.h"
#include "minifi-cpp/utils/gsl.h"
#include "utils/ProcessorConfigUtils.h"
#include "minifi-cpp/core/ProcessorDescriptor.h"

namespace org::apache::nifi::minifi::processors {
const char *GenerateFlowFile::DATA_FORMAT_TEXT = "Text";

constexpr const char * TEXT_CHARS = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890!@#$%^&*()-_=+/?.,';:\"?<>\n\t ";

void GenerateFlowFile::initialize() {
  setSupportedProperties(Properties);
  setSupportedRelationships(Relationships);
}

void generateData(std::vector<char>& data, const bool text_data = false) {
  std::random_device rd;
  std::mt19937 eng(rd());
  if (text_data) {
    const int index_of_last_char = gsl::narrow<int>(strlen(TEXT_CHARS)) - 1;
    std::uniform_int_distribution<> distr(0, index_of_last_char);
    std::generate_n(data.begin(), data.size(), [&] { return TEXT_CHARS[static_cast<uint8_t>(distr(eng))]; });
  } else {
    std::uniform_int_distribution<> distr(std::numeric_limits<char>::min(), std::numeric_limits<char>::max());
    auto rand = [&distr, &eng] { return distr(eng); };
    std::generate_n(data.begin(), data.size(), rand);
  }
}

GenerateFlowFile::Mode GenerateFlowFile::getMode(bool is_unique, bool is_text, bool has_custom_text, uint64_t file_size) {
  if (is_text && !is_unique && has_custom_text)
    return Mode::CustomText;

  if (file_size == 0)
    return Mode::Empty;

  if (is_unique) {
    if (is_text)
      return Mode::UniqueText;
    else
      return Mode::UniqueByte;
  } else {
    if (is_text)
      return Mode::NotUniqueText;
    else
      return Mode::NotUniqueByte;
  }
}

void GenerateFlowFile::onSchedule(core::ProcessContext& context, core::ProcessSessionFactory&) {
  bool is_text = context.getProperty(DataFormat)
      | utils::transform([](const std::string& data_format) { return data_format == DATA_FORMAT_TEXT;})
      | utils::valueOrElse([]() {return false;});
  bool is_unique = utils::parseOptionalBoolProperty(context, UniqueFlowFiles).value_or(true);

  const bool has_custom_text = context.hasNonEmptyProperty(CustomText.name);

  file_size_ = utils::parseDataSizeProperty(context, FileSize);
  batch_size_ = utils::parseU64Property(context, BatchSize);

  mode_ = getMode(is_unique, is_text, has_custom_text, file_size_);

  if (!isUnique(mode_)) {
    non_unique_data_.resize(gsl::narrow<size_t>(file_size_));
    generateData(non_unique_data_, isText(mode_));
  }

  logger_->log_trace("GenerateFlowFile is configured in {} mode", magic_enum::enum_name(mode_));
  if (mode_ != Mode::CustomText && has_custom_text)
    logger_->log_warn("Custom Text property is set but not used. "
      "For Custom Text to be used, Data Format needs to be Text, and Unique FlowFiles needs to be false.");
}

// The custom text has to be reevaluated once per batch
void GenerateFlowFile::refreshNonUniqueData(core::ProcessContext& context) {
  if (mode_ != Mode::CustomText)
    return;
  std::string custom_text = context.getProperty(CustomText).value_or("");
  non_unique_data_.assign(custom_text.begin(), custom_text.end());
}

void GenerateFlowFile::onTrigger(core::ProcessContext& context, core::ProcessSession& session) {
  refreshNonUniqueData(context);
  for (uint64_t i = 0; i < batch_size_; i++) {
    std::shared_ptr<core::FlowFile> flow_file = session.create();
    if (!flow_file) {
      logger_->log_error("Failed to create flowfile!");
      return;
    }
    if (mode_ == Mode::Empty) {
      // the newly created flowfile is empty by default
    } else if (isUnique(mode_)) {
      std::vector<char> unique_data(gsl::narrow<size_t>(file_size_));
      generateData(unique_data, isText(mode_));
      session.writeBuffer(flow_file, unique_data);
    } else {
      session.writeBuffer(flow_file, non_unique_data_);
    }
    session.transfer(flow_file, Success);
  }
}

REGISTER_RESOURCE(GenerateFlowFile, Processor);

}  // namespace org::apache::nifi::minifi::processors

File: minifi_rust/extensions/minifi_rs_playground/src/processors/generate_flow_file.rs
use minifi_native::macros::{ComponentIdentifier, NoAdvancedProcessorFeatures};
use minifi_native::{
    CalculateMetrics, GetProperty, Logger, MinifiError, OnTriggerResult, ProcessContext,
    ProcessSession, Schedule, Trigger,
};
use rand::RngExt;
use rand::distr::Alphanumeric;
use std::cmp::PartialEq;

mod properties;
mod relationships;

#[derive(Debug, PartialEq)]
enum Mode {
    UniqueBytes,
    UniqueText,
    NotUniqueBytes,
    NotUniqueText,
    CustomText,
    Empty,
}

#[derive(Debug, ComponentIdentifier, NoAdvancedProcessorFeatures)]
pub(crate) struct GenerateFlowFileRs {
    mode: Mode,
    batch_size: u64,
    file_size: u64,
    data_generated_during_on_schedule: Vec<u8>,
}

impl Schedule for GenerateFlowFileRs {
    fn schedule<P: GetProperty, L: Logger>(context: &P, _logger: &L) -> Result<Self, MinifiError>
    where
        Self: Sized,
    {
        let is_unique = context
            .get_bool_property(&properties::UNIQUE_FLOW_FILES)?
            .expect("Required property");
        let is_text = context
            .get_property(&properties::DATA_FORMAT)?
            .expect("Required property")
            .as_str()
            == "Text";
        let has_custom_text = context.get_property(&properties::CUSTOM_TEXT)?.is_some();

        let file_size = context
            .get_size_property(&properties::FILE_SIZE)?
            .expect("Required property");
        let batch_size = context
            .get_u64_property(&properties::BATCH_SIZE)?
            .expect("Required property");

        let mode = Self::get_mode(is_unique, is_text, has_custom_text, file_size);
        let data_generated_during_on_schedule =
            if mode == Mode::NotUniqueText || mode == Mode::NotUniqueBytes {
                let mut data = vec![0; file_size as usize];
                Self::generate_data(&mut data, is_text);
                data
            } else {
                vec![]
            };

        Ok(Self {
            mode,
            batch_size,
            file_size,
            data_generated_during_on_schedule,
        })
    }
}

impl GenerateFlowFileRs {
    fn is_unique(&self) -> bool {
        match self.mode {
            Mode::UniqueBytes => true,
            Mode::UniqueText => true,
            Mode::NotUniqueBytes => false,
            Mode::NotUniqueText => false,
            Mode::CustomText => false,
            Mode::Empty => false,
        }
    }

    fn is_text(&self) -> bool {
        match self.mode {
            Mode::UniqueBytes => false,
            Mode::UniqueText => true,
            Mode::NotUniqueBytes => false,
            Mode::NotUniqueText => true,
            Mode::CustomText => true,
            Mode::Empty => false,
        }
    }

    fn get_mode(is_unique: bool, is_text: bool, has_custom_text: bool, file_size: u64) -> Mode {
        if is_text && !is_unique && has_custom_text {
            return Mode::CustomText;
        }

        if file_size == 0 {
            return Mode::Empty;
        }

        match (is_unique, is_text) {
            (true, true) => Mode::UniqueText,
            (true, false) => Mode::UniqueBytes,
            (false, true) => Mode::NotUniqueText,
            (false, false) => Mode::NotUniqueBytes,
        }
    }

    fn generate_data(data: &mut [u8], text_data: bool) {
        let mut rng = rand::rng();

        if text_data {
            for byte in data.iter_mut() {
                *byte = rng.sample(Alphanumeric);
            }
        } else {
            rng.fill(data);
        }
    }
}

impl Trigger for GenerateFlowFileRs {
    fn trigger<PC, PS, L>(
        &self,
        context: &mut PC,
        session: &mut PS,
        _logger: &L,
    ) -> Result<OnTriggerResult, MinifiError>
    where
        PC: ProcessContext,
        PS: ProcessSession<FlowFile = PC::FlowFile>,
    {
        let non_unique_data_buffer: &[u8];
        let custom_text_for_batch: Option<String>;

        if self.mode == Mode::CustomText
            && let Some(custom_text) = context.get_property(&properties::CUSTOM_TEXT, None)?
        {
            custom_text_for_batch = Some(custom_text);
            non_unique_data_buffer = custom_text_for_batch.as_ref().unwrap().as_bytes();
        } else {
            non_unique_data_buffer = self.data_generated_during_on_schedule.as_slice();
        }

        for _ in 0..self.batch_size {
            let mut ff = session.create()?;
            if self.mode != Mode::Empty {
                if self.is_unique() {
                    let mut unique_data: Vec<u8> = vec![0; self.file_size as usize];
                    Self::generate_data(&mut unique_data, self.is_text());
                    session.write(&mut ff, unique_data.as_slice())?;
                } else {
                    session.write(&mut ff, non_unique_data_buffer)?;
                }
            }
            session.transfer(ff, relationships::SUCCESS.name)?;
        }
        Ok(OnTriggerResult::Ok)
    }
}

impl CalculateMetrics for GenerateFlowFileRs {}

pub(crate) mod processor_definition;

#[cfg(test)]
mod tests;
